File: csharp/DVSClient/DVSClient/address/format/AddressFormatted.cs
using DVSClient.Server.Address.Format;

namespace DVSClient.Address.Format
{
    public class AddressFormatted
    {
        internal string LayoutName { get; }
        internal bool? HasEnoughLines { get; }
        internal bool? HasTruncatedLines { get; }
        internal IDictionary<string, string> Address { get; }
        internal bool? HasMissingSubPremises { get; }
        internal IEnumerable<AddressFormattedLine> AddressLines { get; }

        public AddressFormatted(RestApiAddressFormatted result)
        {
            LayoutName = result.LayoutName ?? string.Empty;
            HasEnoughLines = result.NotEnoughLines.HasValue ? !result.NotEnoughLines.Value : null;
            HasTruncatedLines = result.HasTruncatedLines;
            Address = new Dictionary<string, string>();
            if (result.Address != null)
            {
                foreach (var entry in result.Address)
                {
                    Address[entry.Key] = entry.Value ?? string.Empty;
                }
            }
            HasMissingSubPremises = result.HasMissingSubPremises;
            AddressLines = result.AddressLines != null ? result.AddressLines.ToList().ConvertAll(line => new AddressFormattedLine(line)) : new List<AddressFormattedLine>();
        }
    }
}

File: typescript/src/address/format/addressFormatted.ts
import { RestApiAddressFormatted } from "../../server/address/format/restApiAddressFormatted";
import { AddressFormattedLine, restApiToAddressFormattedLine } from "./addressFormattedLine";

export type AddressFormatted = {
    notEnoughLines: boolean;
    layoutName: string;
    hasTruncatedLines: boolean;
    address: { [key: string]: string };
    hasMissingSubPremises: boolean;
    addressLines: AddressFormattedLine[];
};

export function restApiToAddressFormatted(response: RestApiAddressFormatted): AddressFormatted {
    
    return {
        layoutName: response.layout_name ?? "",
        notEnoughLines: response.not_enough_lines ?? false,
        hasTruncatedLines: response.has_truncated_lines ?? false,
        address: Object.fromEntries(
            Object.entries(response.address ?? {}).map(([key, value]) => [key, value ?? ""])
        ),
        hasMissingSubPremises: response.has_missing_sub_premises ?? false,
        addressLines: response.address_lines ? response.address_lines.map(line => restApiToAddressFormattedLine(line)) : []
    };
    
}